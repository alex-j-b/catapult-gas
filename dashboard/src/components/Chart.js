// React
import React, { Component } from 'react';
// Libs
import _ from 'lodash';
import { ResponsiveLine } from '@nivo/line';

// Slice desired part (each or all or d1...)
const sliceData = (d, m) => {
    const params =  m.params.option ? m.params : { option: 'all' };
    const sliceOptions = {
        all: [0, 1],
        each: [1, d.length],
        d: [Number(params.deviceId), Number(params.deviceId) + 1]
    };
    const target = sliceOptions[params.option];
    const slicedData = d.slice(target[0], target[1]);

    return slicedData;
}


class Chart extends Component {
    state = {
        slicedData: sliceData(this.props.data, this.props.match)
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        const newSlice = sliceData(nextProps.data, nextProps.match);
        // Update total liters display
        const totalDisplay = newSlice.length > 1 ? nextProps.data[0].total : newSlice[0].total;
        nextProps.setTotal(totalDisplay.toLocaleString());

        return {
            slicedData: newSlice
        }
    }

    shouldComponentUpdate(nextProps, nextState) {
        return !_.isEqual(this.props.match.params, nextProps.match.params);
    }

    componentDidMount() {
        // Update total liters display
        const totalDisplay = this.state.slicedData.length > 1 ? this.props.data[0].total : this.state.slicedData[0].total;
        this.props.setTotal(totalDisplay.toLocaleString());
    }

    render() {
        return (
            <ResponsiveLine
                data={this.state.slicedData}
                margin={{ top: 60, right: 140, bottom: 50, left: 80 }}
                theme={{
                    fontSize: 12,
                    axis: { legend: { text: { fontSize: 16 } } }
                }}
                xScale={{ type: "time", format: "%d-%m-%Y" }}
                xFormat="time:%d-%m-%Y"
                axisBottom={{
                    format: "%d-%m-%Y",
                    orient: 'bottom',
                    tickSize: 5,
                    tickPadding: 10
                }}
                yScale={{ type: 'linear', min: 0, max: 'auto', stacked: false, reverse: false }}
                yFormat=" >-.2f"
                axisLeft={{
                    orient: 'left',
                    tickSize: 5,
                    tickPadding: 10,
                    tickRotation: 0
                }}
                pointSize={10}
                pointColor={{ theme: 'background' }}
                pointBorderWidth={2}
                pointBorderColor={{ from: 'serieColor' }}
                pointLabelYOffset={-12}
                useMesh={true}
                legends={[
                    {
                        anchor: 'bottom-right',
                        direction: 'column',
                        justify: false,
                        translateX: 100,
                        translateY: 0,
                        itemsSpacing: 0,
                        itemDirection: 'left-to-right',
                        itemWidth: 50,
                        itemHeight: 40,
                        itemOpacity: 0.75,
                        symbolSize: 20,
                        symbolShape: 'circle'
                    }
                ]}
            />
        )
    }
}

export default Chart;
