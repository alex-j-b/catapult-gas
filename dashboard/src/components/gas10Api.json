[{
    "id": "all",
    "total": 3510224,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 495686
    }, {
        "x": "17-09-2021",
        "y": 468478
    }, {
        "x": "16-09-2021",
        "y": 432215
    }, {
        "x": "15-09-2021",
        "y": 407038
    }, {
        "x": "14-09-2021",
        "y": 399147
    }, {
        "x": "13-09-2021",
        "y": 378107
    }, {
        "x": "12-09-2021",
        "y": 334569
    }, {
        "x": "11-09-2021",
        "y": 317297
    }, {
        "x": "10-09-2021",
        "y": 277687
    }]
}, {
    "id": "1",
    "total": 260174,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 37112
    }, {
        "x": "17-09-2021",
        "y": 32184
    }, {
        "x": "16-09-2021",
        "y": 29322
    }, {
        "x": "15-09-2021",
        "y": 29641
    }, {
        "x": "14-09-2021",
        "y": 29293
    }, {
        "x": "13-09-2021",
        "y": 26757
    }, {
        "x": "12-09-2021",
        "y": 26983
    }, {
        "x": "11-09-2021",
        "y": 26208
    }, {
        "x": "10-09-2021",
        "y": 22674
    }]
}, {
    "id": "2",
    "total": 482646,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 71454
    }, {
        "x": "17-09-2021",
        "y": 63467
    }, {
        "x": "16-09-2021",
        "y": 57797
    }, {
        "x": "15-09-2021",
        "y": 53545
    }, {
        "x": "14-09-2021",
        "y": 56537
    }, {
        "x": "13-09-2021",
        "y": 51222
    }, {
        "x": "12-09-2021",
        "y": 46387
    }, {
        "x": "11-09-2021",
        "y": 44363
    }, {
        "x": "10-09-2021",
        "y": 37874
    }]
}, {
    "id": "3",
    "total": 692519,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 95914
    }, {
        "x": "17-09-2021",
        "y": 93757
    }, {
        "x": "16-09-2021",
        "y": 86887
    }, {
        "x": "15-09-2021",
        "y": 77890
    }, {
        "x": "14-09-2021",
        "y": 78088
    }, {
        "x": "13-09-2021",
        "y": 76291
    }, {
        "x": "12-09-2021",
        "y": 67304
    }, {
        "x": "11-09-2021",
        "y": 60394
    }, {
        "x": "10-09-2021",
        "y": 55994
    }]
}, {
    "id": "4",
    "total": 918286,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 129064
    }, {
        "x": "17-09-2021",
        "y": 122029
    }, {
        "x": "16-09-2021",
        "y": 114831
    }, {
        "x": "15-09-2021",
        "y": 112590
    }, {
        "x": "14-09-2021",
        "y": 102063
    }, {
        "x": "13-09-2021",
        "y": 103962
    }, {
        "x": "12-09-2021",
        "y": 81983
    }, {
        "x": "11-09-2021",
        "y": 77929
    }, {
        "x": "10-09-2021",
        "y": 73835
    }]
}, {
    "id": "5",
    "total": 1156599,
    "data": [{
        "x": "19-09-2021",
        "y": 0
    }, {
        "x": "18-09-2021",
        "y": 162142
    }, {
        "x": "17-09-2021",
        "y": 157041
    }, {
        "x": "16-09-2021",
        "y": 143378
    }, {
        "x": "15-09-2021",
        "y": 133372
    }, {
        "x": "14-09-2021",
        "y": 133166
    }, {
        "x": "13-09-2021",
        "y": 119875
    }, {
        "x": "12-09-2021",
        "y": 111912
    }, {
        "x": "11-09-2021",
        "y": 108403
    }, {
        "x": "10-09-2021",
        "y": 87310
    }]
}]
