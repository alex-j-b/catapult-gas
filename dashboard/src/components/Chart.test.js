import Enzyme, { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Chart from './Chart';
import gas10Api from './gas10Api.json';
Enzyme.configure({ adapter: new Adapter() });

describe('Chart tests', () => {
    it('should render chart correctly', () => {
        const match = {
            path: '/',
            url: '/',
            isExact: true,
            params: {}
        };

        const component = shallow(
            <Chart match={match} data={gas10Api} setTotal={(e) => { return null}} />
        );

        expect(toJSON(component)).toMatchSnapshot();
    });
});
