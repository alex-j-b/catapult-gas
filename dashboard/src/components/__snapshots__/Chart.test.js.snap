// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart tests should render chart correctly 1`] = `
<ResponsiveLine
  axisBottom={
    Object {
      "format": "%d-%m-%Y",
      "orient": "bottom",
      "tickPadding": 10,
      "tickSize": 5,
    }
  }
  axisLeft={
    Object {
      "orient": "left",
      "tickPadding": 10,
      "tickRotation": 0,
      "tickSize": 5,
    }
  }
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "x": "19-09-2021",
            "y": 0,
          },
          Object {
            "x": "18-09-2021",
            "y": 495686,
          },
          Object {
            "x": "17-09-2021",
            "y": 468478,
          },
          Object {
            "x": "16-09-2021",
            "y": 432215,
          },
          Object {
            "x": "15-09-2021",
            "y": 407038,
          },
          Object {
            "x": "14-09-2021",
            "y": 399147,
          },
          Object {
            "x": "13-09-2021",
            "y": 378107,
          },
          Object {
            "x": "12-09-2021",
            "y": 334569,
          },
          Object {
            "x": "11-09-2021",
            "y": 317297,
          },
          Object {
            "x": "10-09-2021",
            "y": 277687,
          },
        ],
        "id": "all",
        "total": 3510224,
      },
    ]
  }
  legends={
    Array [
      Object {
        "anchor": "bottom-right",
        "direction": "column",
        "itemDirection": "left-to-right",
        "itemHeight": 40,
        "itemOpacity": 0.75,
        "itemWidth": 50,
        "itemsSpacing": 0,
        "justify": false,
        "symbolShape": "circle",
        "symbolSize": 20,
        "translateX": 100,
        "translateY": 0,
      },
    ]
  }
  margin={
    Object {
      "bottom": 50,
      "left": 80,
      "right": 140,
      "top": 60,
    }
  }
  pointBorderColor={
    Object {
      "from": "serieColor",
    }
  }
  pointBorderWidth={2}
  pointColor={
    Object {
      "theme": "background",
    }
  }
  pointLabelYOffset={-12}
  pointSize={10}
  theme={
    Object {
      "axis": Object {
        "legend": Object {
          "text": Object {
            "fontSize": 16,
          },
        },
      },
      "fontSize": 12,
    }
  }
  useMesh={true}
  xFormat="time:%d-%m-%Y"
  xScale={
    Object {
      "format": "%d-%m-%Y",
      "type": "time",
    }
  }
  yFormat=" >-.2f"
  yScale={
    Object {
      "max": "auto",
      "min": 0,
      "reverse": false,
      "stacked": false,
      "type": "linear",
    }
  }
/>
`;
